packages/core/src/services/ai/index.ts

```diff
@@ -57,159 +63,99 @@ export async function ai({
   schema,
   output,
   customLanguageModel,
+  aiSdkProvider,
 }: {
   provider: ProviderApiKey
   config: PartialConfig
   messages: Message[]
   prompt?: string
   schema?: JSONSchema7
-  output?: 'object' | 'array' | 'no-schema' | undefined
   customLanguageModel?: LanguageModel
+  output?: ObjectOutput
+  aiSdkProvider?: Partial<AISDKProvider>
 }): Promise<
   TypedResult<
     AIReturn<StreamType>,
-    ChainError<RunErrorCodes.AIProviderConfigError>
+    ChainError<
+      | RunErrorCodes.AIProviderConfigError
+      | RunErrorCodes.AIRunError
+      | RunErrorCodes.Unknown
+    >
   >
 > {
-  const { provider, token: apiKey, url } = apiProvider
-  const model = config.model
+  const { streamText, streamObject } = {
+    ...DEFAULT_AI_SDK_PROVIDER,
+    ...(aiSdkProvider || {}),
+  }
+  try {
+    const { provider, token: apiKey, url } = apiProvider
+    const model = config.model
+
+    const languageModelResult = createProvider({
+      messages,
+      provider,
+      apiKey,
+      config,
+      ...(url ? { url } : {}),
+    })
 
-  const languageModelResult = createProvider({
-    messages,
-    provider,
-    apiKey,
-    config,
-    ...(url ? { url } : {}),
-  })
+    if (languageModelResult.error) return languageModelResult
 
-  if (languageModelResult.error) return languageModelResult
+    const languageModel = customLanguageModel
+      ? customLanguageModel
+      : languageModelResult.value(model)
+    const toolsResult = buildTools(config.tools)
+    if (toolsResult.error) return toolsResult
 
-  const languageModel = customLanguageModel
-    ? customLanguageModel
-    : languageModelResult.value(model)
-  const toolsResult = buildTools(config.tools)
-  if (toolsResult.error) return toolsResult
+    const commonOptions = {
+      ...omit(config, ['schema']),
+      model: languageModel,
+      prompt,
+      messages: messages as CoreMessage[],
+      tools: toolsResult.value,
+    }
 
-  const commonOptions = {
-    ...omit(config, ['schema']),
-    model: languageModel,
-    prompt,
-    messages: messages as CoreMessage[],
-    tools: toolsResult.value,
-  }
+    if (UNSUPPORTED_STREAM_MODELS.includes(model)) {
+      return runNoStreamingModels({
+        schema,
+        output,
+        commonOptions,
+      })
+    }
 
-  if (UNSUPPORTED_STREAM_MODELS.includes(model)) {
-    if (output && schema) {
```

Comment:
2 if level nesting + the try/catch now. I moved non-streaming to `runNoStreamingModels` service